jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/component/Environment.java

```diff
@@ -13,41 +13,68 @@
 
 package org.eclipse.jetty.util.component;
 
-import java.io.IOException;
 import java.util.Collection;
 import java.util.Collections;
-import java.util.Map;
-import java.util.TreeMap;
 
 import org.eclipse.jetty.util.Attributes;
-import org.eclipse.jetty.util.TypeUtil;
 
 /**
  * A named runtime environment containing a {@link ClassLoader} and {@link Attributes}.
  */
 public interface Environment extends Attributes
 {
     // Ensure there is a core environment for possible later deployments to it
-    Environment CORE = ensure("core");
+    Environment CORE = ensure("core", Environment.class.getClassLoader());
 
+    /**
+     * Gets all existing environments.
+     * @return the environments
+     */
     static Collection<Environment> getAll()
     {
-        return Collections.unmodifiableCollection(Named.__environments.values());
+        return Collections.unmodifiableCollection(NamedEnvironment.ENVIRONMENTS.values());
     }
-    
+
+    /**
+     * Gets the environment with the given name.
+     * @param name the environment name
+     * @return the environment, or null if no environment with such name exists
+     */
     static Environment get(String name)
     {
-        return Named.__environments.get(name);
+        return NamedEnvironment.ENVIRONMENTS.get(name);
     }
 
-    static Environment ensure(String name)
+    /**
+     * Gets the environment with the given name, creating it with the default classloader if necessary.
+     *
+     * @param name the environment name
+     * @return the environment
+     * @throws IllegalStateException if an environment with the given name but a non-default classloader already exists
+     */
+    static Environment ensure(String name, ClassLoader classLoader) throws IllegalStateException
     {
-        return Named.__environments.computeIfAbsent(name, Named::new);
+        Environment environment = NamedEnvironment.ENVIRONMENTS.computeIfAbsent(name, n -> new NamedEnvironment(n, classLoader));
+        if (environment.getClassLoader() != classLoader)
```

Comment:
So this check means that `EnvironmentBuilder` and `XmlConfiguration` must not configure `ee*` environments otherwise the `ensure()` call in the multiple `ServletContextHandler`s would make those classes fail to load because of this check.

Did I get this right?

Reply:
Not at all.   The EB and XC can configure ee* environments, so long as they do so in a way that classes that call ensure are loaded from the classloaders in the created environments

Reply:
ok, thanks for the explanation.